[standalone_compile] Some misc fixes

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #151502
* #151551
* #151501

This PR fixes two things.

The first problem is that in the vLLM style standalone_compile is
called from within a custom torch.compile backend. If there already is a
FakeTensorMode (which there is), we shouldn't create a new
FakeTensorMode with the same shape_env, instead we should just reuse the
same FakeTensorMode.

The second thing is that compile_fx can mutate the passed in gm, so we
deepcopy (since standalone_compile should be standalone)

Test Plan:
- new test
- updated old tests

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

zou3519 committed:
[standalone_compile] Some misc fixes

This PR fixes two things. They're a bit difficult to separate so I put
them together.

The first problem is handling multiple returns. standalone_compile is
supposed to replicate the calling convention of the graph module passed
to it. If the graph module returns a tuple of a single Tensor, then
standalone_compile should return a function that returns a tuple of a
single Tensor. This was a simple fix, just removing the [0], and this
allows us to handle multiple returns in the function being compiled too.

The second problem is that in the vLLM style standalone_compile is
called from within a custom torch.compile backend. If there already is a
FakeTensorMode (which there is), we shouldn't create a new
FakeTensorMode with the same shape_env, instead we should just reuse the
same FakeTensorMode.

Test Plan:
- new test
- updated old tests

[ghstack-poisoned]

zou3519 committed:
Update on "[standalone_compile] Some misc fixes"


This PR fixes two things. They're a bit difficult to separate so I put
them together.

The first problem is handling multiple returns. standalone_compile is
supposed to replicate the calling convention of the graph module passed
to it. If the graph module returns a tuple of a single Tensor, then
standalone_compile should return a function that returns a tuple of a
single Tensor. This was a simple fix, just removing the [0], and this
allows us to handle multiple returns in the function being compiled too.

The second problem is that in the vLLM style standalone_compile is
called from within a custom torch.compile backend. If there already is a
FakeTensorMode (which there is), we shouldn't create a new
FakeTensorMode with the same shape_env, instead we should just reuse the
same FakeTensorMode.

Test Plan:
- new test
- updated old tests

[ghstack-poisoned]

zou3519 committed:
Update on "[standalone_compile] Some misc fixes"


This PR fixes three things. They're a bit difficult to separate so I put
them together.

The first problem is handling multiple returns. standalone_compile is
supposed to replicate the calling convention of the graph module passed
to it. If the graph module returns a tuple of a single Tensor, then
standalone_compile should return a function that returns a tuple of a
single Tensor. This was a simple fix, just removing the [0], and this
allows us to handle multiple returns in the function being compiled too.

The second problem is that in the vLLM style standalone_compile is
called from within a custom torch.compile backend. If there already is a
FakeTensorMode (which there is), we shouldn't create a new
FakeTensorMode with the same shape_env, instead we should just reuse the
same FakeTensorMode.

The third thing is that compile_fx can mutate the passed in gm, so we
deepcopy (since standalone_compile should be standalone)

Test Plan:
- new test
- updated old tests

[ghstack-poisoned]

zou3519 committed:
Update on "[standalone_compile] Some misc fixes"


This PR fixes three things. They're a bit difficult to separate so I put
them together.

The first problem is handling multiple returns. standalone_compile is
supposed to replicate the calling convention of the graph module passed
to it. If the graph module returns a tuple of a single Tensor, then
standalone_compile should return a function that returns a tuple of a
single Tensor. This was a simple fix, just removing the [0], and this
allows us to handle multiple returns in the function being compiled too.

The second problem is that in the vLLM style standalone_compile is
called from within a custom torch.compile backend. If there already is a
FakeTensorMode (which there is), we shouldn't create a new
FakeTensorMode with the same shape_env, instead we should just reuse the
same FakeTensorMode.

The third thing is that compile_fx can mutate the passed in gm, so we
deepcopy (since standalone_compile should be standalone)

Test Plan:
- new test
- updated old tests

[ghstack-poisoned]

zou3519 committed:
Update on "[standalone_compile] Some misc fixes"


This PR fixes two things.

The first problem is that in the vLLM style standalone_compile is
called from within a custom torch.compile backend. If there already is a
FakeTensorMode (which there is), we shouldn't create a new
FakeTensorMode with the same shape_env, instead we should just reuse the
same FakeTensorMode.

The second thing is that compile_fx can mutate the passed in gm, so we
deepcopy (since standalone_compile should be standalone)

Test Plan:
- new test
- updated old tests

[ghstack-poisoned]

oulgen commented on torch/_inductor/standalone_compile.py:
```diff
@@ -166,12 +166,15 @@ def standalone_compile(
 
     from .compile_fx import compile_fx
 
-    shape_env = shape_env_from_inputs(example_inputs, default=True)
```
i guess you can revert my changes in this function too

zou3519 replied:
yup will do

zou3519 committed:
Update on "[standalone_compile] Some misc fixes"


This PR fixes two things.

The first problem is that in the vLLM style standalone_compile is
called from within a custom torch.compile backend. If there already is a
FakeTensorMode (which there is), we shouldn't create a new
FakeTensorMode with the same shape_env, instead we should just reuse the
same FakeTensorMode.

The second thing is that compile_fx can mutate the passed in gm, so we
deepcopy (since standalone_compile should be standalone)

Test Plan:
- new test
- updated old tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
